pkg/cli/delete/progress.go

```diff
@@ -0,0 +1,170 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package delete
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/gosuri/uilive"
+	"github.com/mattn/go-isatty"
+	"github.com/radius-project/radius/pkg/cli/clients"
+	"github.com/radius-project/radius/pkg/cli/output"
+)
+
+// NewProgressListener creates a new ProgressListener based on whether the output is a terminal or not, returning an
+// InteractiveListener if it is a terminal and a NoOpListener if it is not.
+func NewProgressListener(progressChan <-chan clients.ResourceProgress) ProgressListener {
+	if isatty.IsTerminal(os.Stdout.Fd()) {
+		return &InteractiveListener{
+			progressChan: progressChan,
+			writerDone:   &sync.WaitGroup{},
+			Spinner:      output.ProgressDefaultSpinner,
+		}
+	} else {
+		return &NoOpListener{
+			progressChan: progressChan,
+		}
+	}
+}
+
+type ProgressListener interface {
+	// Run is called to print progress to the command line. This should be called from
+	// a goroutine because it blocks until the progress channel is closed.
+	Run()
+}
+
+type NoOpListener struct {
+	progressChan <-chan clients.ResourceProgress
+}
+
+// Run() continuously drains the updates from the progressChan channel without taking any action.
+func (listener *NoOpListener) Run() {
+	for range listener.progressChan {
+		// Do nothing except drain the updates.
+	}
+}
+
+type InteractiveListener struct {
+	progressChan <-chan clients.ResourceProgress
+	Spinner      []string
+	mutex        sync.Mutex
+	entries      []Entry
+	spinnerIndex int
+	writerDone   *sync.WaitGroup
+}
+
+type Entry struct {
+	// FinalState is an optional token that will replace the spinner with static text.
+	FinalState string
+
+	// Format is the format string used to build the output line. It is expected to contain a placeholder
+	// for the spinner/final-state token.
+	Format string
+}
+
+func (listener *InteractiveListener) addEntry(format string) int {
+	listener.mutex.Lock()
+	defer listener.mutex.Unlock()
+
+	listener.entries = append(listener.entries, Entry{Format: format})
+	return len(listener.entries) - 1
+}
+
+func (listener *InteractiveListener) updateEntry(index int, state string, format string) {
+	listener.mutex.Lock()
+	defer listener.mutex.Unlock()
+
+	listener.entries[index] = Entry{FinalState: state, Format: format}
+}
+
+// Run() concurrently updates the UI with a spinner and writes output for each resource update received from the progressChan channel.
+func (listener *InteractiveListener) Run() {
```

Comment:
The goroutine doesn't have any error recovery mechanism. Consider adding a defer with recover() to prevent panics from crashing the application.